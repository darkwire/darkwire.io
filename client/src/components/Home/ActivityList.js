import React from 'react'
import PropTypes from 'prop-types'
import ChatInput from 'components/Chat'
import Activity from './Activity'
import T from 'components/T'
import { defer } from 'lodash'

import styles from './styles.module.scss'

const ActivityList = ({ activities, openModal }) => {
  const [focusChat, setFocusChat] = React.useState(false);
  const [scrolledToBottom, setScrolledToBottom] = React.useState(true);
  const messageStream = React.useRef(null);
  const activitiesList = React.useRef(null);

  React.useEffect(() => {
    const currentMessageStream = messageStream.current;

    // Update scrolledToBottom state if we scroll the activity stream
    const onScroll = () => {
      const messageStreamHeight = messageStream.current.clientHeight
      const activitiesListHeight = activitiesList.current.clientHeight

      const bodyRect = document.body.getBoundingClientRect()
      const elemRect = activitiesList.current.getBoundingClientRect()
      const offset = elemRect.top - bodyRect.top
      const activitiesListYPos = offset

      const newScrolledToBottom = (activitiesListHeight + (activitiesListYPos - 60)) <= messageStreamHeight
      if (newScrolledToBottom) {
        if (!scrolledToBottom) {
          setScrolledToBottom(true)
        }
      } else if (scrolledToBottom) {
        setScrolledToBottom(false)
      }
    }

    currentMessageStream.addEventListener('scroll', onScroll)
    return () => {
      // Unbind event if component unmounted
      currentMessageStream.removeEventListener('scroll', onScroll)
    }
  }, [scrolledToBottom])

  const scrollToBottomIfShould = React.useCallback(() => {
    if (scrolledToBottom) {
      messageStream.current.scrollTop = messageStream.current.scrollHeight
    }
  }, [scrolledToBottom])

  React.useEffect(() => {
    scrollToBottomIfShould(); // Only if activities.length bigger
  }, [scrollToBottomIfShould, activities]);

  const scrollToBottom = React.useCallback(() => {
    messageStream.current.scrollTop = messageStream.current.scrollHeight
    setScrolledToBottom(true)
  }, [])

  const handleChatClick = () => {
    setFocusChat(true);
    defer(() => setFocusChat(false))
  }

  return (
    <div className="main-chat">
      <div onClick={handleChatClick} className="message-stream h-100" ref={messageStream} data-testid="main-div">
        <ul className="plain" ref={activitiesList}>
          <li><p className={styles.tos}><button className='btn btn-link' onClick={() => openModal('About')}> <T path='agreement'/></button></p></li>
          {activities.map((activity, index) => (
            <li key={index} className={`activity-item ${activity.type}`}>
              <Activity activity={activity} scrollToBottom={scrollToBottomIfShould} />
            </li>
          ))}
        </ul>
      </div>
      <div className="chat-container">
        <ChatInput scrollToBottom={scrollToBottom} focusChat={focusChat} />
      </div>
    </div>
  )
}

ActivityList.propTypes = {
  activities: PropTypes.array.isRequired,
  openModal: PropTypes.func.isRequired,
}

export default ActivityList
