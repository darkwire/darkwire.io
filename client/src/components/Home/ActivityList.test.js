import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import ActivityList from './ActivityList';
import { Provider } from 'react-redux';
import configureStore from 'store';

const store = configureStore();

jest.useFakeTimers();

describe('ActivityList component', () => {
  it('should display', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <ActivityList openModal={jest.fn()} activities={[]} setScrolledToBottom={jest.fn()} scrolledToBottom />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display with activities', () => {
    const activities = [
      {
        type: 'TEXT_MESSAGE',
        username: 'alice',
        timestamp: new Date('2020-03-14T11:01:58.135Z').valueOf(),
        text: 'Hi!',
      },
      {
        type: 'USER_ENTER',
        username: 'alice',
      },
      {
        type: 'CHANGE_USERNAME',
        currentUsername: 'alice',
        newUsername: 'alicette',
      },
    ];
    const { asFragment } = render(
      <Provider store={store}>
        <ActivityList openModal={jest.fn()} activities={activities} setScrolledToBottom={jest.fn()} scrolledToBottom />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should show About modal', async () => {
    const mockOpenModal = jest.fn();

    const { getByText } = render(
      <Provider store={store}>
        <ActivityList openModal={mockOpenModal} activities={[]} setScrolledToBottom={jest.fn()} scrolledToBottom />
      </Provider>,
    );

    fireEvent.click(getByText('By using Darkwire, you are agreeing to our Acceptable Use Policy and Terms of Service'));
    jest.runAllTimers();

    expect(mockOpenModal.mock.calls[0][0]).toBe('About');
  });

  it('should focus chat', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <ActivityList openModal={jest.fn()} activities={[]} setScrolledToBottom={jest.fn()} scrolledToBottom />
      </Provider>,
    );
    fireEvent.click(getByTestId('main-div'));
    jest.runAllTimers();
  });

  it('should handle scroll', () => {
    const mockSetScrollToBottom = jest.fn();
    jest
      .spyOn(Element.prototype, 'clientHeight', 'get')
      .mockReturnValueOnce(400)
      .mockReturnValueOnce(200)
      .mockReturnValueOnce(400)
      .mockReturnValueOnce(200);

    Element.prototype.getBoundingClientRect = jest
      .fn()
      .mockReturnValueOnce({ top: 0 })
      .mockReturnValueOnce({ top: 60 })
      .mockReturnValueOnce({ top: 0 })
      .mockReturnValueOnce({ top: 261 });

    const { getByTestId, rerender } = render(
      <Provider store={store}>
        <ActivityList
          openModal={jest.fn()}
          activities={[]}
          setScrolledToBottom={mockSetScrollToBottom}
          scrolledToBottom={false}
        />
      </Provider>,
    );
    fireEvent.scroll(getByTestId('main-div'));

    expect(mockSetScrollToBottom).toHaveBeenLastCalledWith(true);

    rerender(
      <Provider store={store}>
        <ActivityList
          openModal={jest.fn()}
          activities={[]}
          setScrolledToBottom={mockSetScrollToBottom}
          scrolledToBottom={true}
        />
      </Provider>,
    );

    fireEvent.scroll(getByTestId('main-div'));

    expect(mockSetScrollToBottom).toHaveBeenCalledTimes(2);
    expect(mockSetScrollToBottom).toHaveBeenLastCalledWith(false);
  });

  it('should scroll to bottom on new message', () => {
    const mockSetScrollToBottom = jest.fn();

    jest.spyOn(Element.prototype, 'scrollHeight', 'get').mockReturnValue(42);
    const mockScrollTop = jest.spyOn(Element.prototype, 'scrollTop', 'set');

    const { rerender } = render(
      <Provider store={store}>
        <ActivityList
          openModal={jest.fn()}
          activities={[]}
          setScrolledToBottom={mockSetScrollToBottom}
          scrolledToBottom={true}
        />
      </Provider>,
    );

    rerender(
      <Provider store={store}>
        <ActivityList
          openModal={jest.fn()}
          activities={[
            {
              type: 'TEXT_MESSAGE',
              username: 'alice',
              timestamp: new Date('2020-03-14T11:01:58.135Z').valueOf(),
              text: 'Hi!',
            },
          ]}
          setScrolledToBottom={mockSetScrollToBottom}
          scrolledToBottom={true}
        />
      </Provider>,
    );

    jest.runAllTimers();

    expect(mockScrollTop).toHaveBeenLastCalledWith(42);
  });
});
