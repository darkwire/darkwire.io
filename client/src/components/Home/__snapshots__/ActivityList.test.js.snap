// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityList component should display 1`] = `
<DocumentFragment>
  <div
    class="main-chat"
  >
    <div
      class="message-stream h-100"
      data-testid="main-div"
    >
      <ul
        class="plain"
      >
        <li>
          <p
            class="tos"
          >
            <button
              class="btn btn-link"
            >
               By using Darkwire, you are agreeing to our Acceptable Use Policy and Terms of Service
            </button>
          </p>
        </li>
      </ul>
    </div>
    <div
      class="chat-container"
    >
      <form
        class="chat-preflight-container"
      >
        <textarea
          class="chat"
          placeholder="Type here"
          rows="1"
        />
        <div
          class="input-controls"
        >
          <div
            class="styles icon file-transfer btn btn-link"
          >
            <input
              id="fileInput"
              name="fileUploader"
              placeholder="Choose a file..."
              type="file"
            />
            <label
              for="fileInput"
            >
              <svg
                fill="none"
                height="24"
                stroke="#fff"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"
                />
                <polyline
                  points="13 2 13 9 20 9"
                />
              </svg>
            </label>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ActivityList component should display with activities 1`] = `
<DocumentFragment>
  <div
    class="main-chat"
  >
    <div
      class="message-stream h-100"
      data-testid="main-div"
    >
      <ul
        class="plain"
      >
        <li>
          <p
            class="tos"
          >
            <button
              class="btn btn-link"
            >
               By using Darkwire, you are agreeing to our Acceptable Use Policy and Terms of Service
            </button>
          </p>
        </li>
        <li
          class="activity-item TEXT_MESSAGE"
        >
          <div>
            <div
              class="chat-meta"
            >
              <span
                class="username"
                style="color: rgb(204, 241, 255);"
              >
                alice
              </span>
              <span
                class="muted timestamp"
              >
                11:01 AM
              </span>
            </div>
            <div
              class="chat"
            >
              <span
                class="Linkify"
              >
                Hi!
              </span>
            </div>
          </div>
        </li>
        <li
          class="activity-item USER_ENTER"
        >
          <div>
            <div
              class="info"
            >
              <div>
                <span>
                  <span
                    class="username"
                    style="color: rgb(204, 241, 255);"
                  >
                    alice
                  </span>
                   joined
                </span>
              </div>
            </div>
          </div>
        </li>
        <li
          class="activity-item CHANGE_USERNAME"
        >
          <div>
            <div
              class="info"
            >
              <div>
                <span>
                  <span
                    class="username"
                    style="color: rgb(204, 241, 255);"
                  >
                    alice
                  </span>
                   changed their name to 
                  <span
                    class="username"
                    style="color: rgb(235, 162, 242);"
                  >
                    alicette
                  </span>
                </span>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="chat-container"
    >
      <form
        class="chat-preflight-container"
      >
        <textarea
          class="chat"
          placeholder="Type here"
          rows="1"
        />
        <div
          class="input-controls"
        >
          <div
            class="styles icon file-transfer btn btn-link"
          >
            <input
              id="fileInput"
              name="fileUploader"
              placeholder="Choose a file..."
              type="file"
            />
            <label
              for="fileInput"
            >
              <svg
                fill="none"
                height="24"
                stroke="#fff"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"
                />
                <polyline
                  points="13 2 13 9 20 9"
                />
              </svg>
            </label>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
